shap/explainers/_tree.py

```diff
@@ -1697,43 +1699,34 @@ class XGBTreeModelLoader:
     tree can actually be wrong when feature values land almost on a threshold.
     """
     def __init__(self, xgb_model):
-        # new in XGBoost 1.1, 'binf' is appended to the buffer
-        self.buf = xgb_model.save_raw()
-        if self.buf.startswith(b'binf'):
-            self.buf = self.buf[4:]
-        self.pos = 0
-
-        # load the model parameters
-        self.base_score = self.read('f')
-        self.num_feature = self.read('I')
-        self.num_class = self.read('i')
-        self.contain_extra_attrs = self.read('i')
-        self.contain_eval_metrics = self.read('i')
-        self.read_arr('i', 29) # reserved
-        self.name_obj_len = self.read('Q')
-        self.name_obj = self.read_str(self.name_obj_len)
-        self.name_gbm_len = self.read('Q')
-        self.name_gbm = self.read_str(self.name_gbm_len)
+        import xgboost
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            if version.parse(xgboost.__version__) >= version.parse("1.6.0"):
+                tmp_file = os.path.join(tmp_dir, "model.ubj")
+                xgb_model.save_model(tmp_file)
+                xgb_params = decode_ubjson_buffer(open(tmp_file, 'rb'))
+            else:
+                tmp_file = os.path.join(tmp_dir, "model.json")
```

Comment:
If the xgboost version is lower than 1.6 (which is the version ubjson was introduced), should we keep the old binary parser?

Reply:
We generally try to follow [SPEC 0](https://scientific-python.org/specs/spec-0000/), which means supporting XBoost versions that were released in the past two years.

From a scan of the XGB releases:
- XGB 1.5.0 was released Oct 2021
- XGB 1.6.0 was released Apr 2022

So, at time of writing I think it would be appropriate for us to drop support for XGB 1.6, if that means this code can be simplified significantly. We should raise an informative error message if the XGB version is not supported.

Reply:
As I understand, at present the code uses the JSON representation for models below 1.6. What are the consequences of that compared to the binary format? Is it worth raising a warning about any limitations?

Reply:
Json is less precise than the binary format. So raising a warning about this is appropriate IMO

Reply:
Done.